core/src/main/scala/com/salesforce/op/stages/impl/insights/RecordInsightsLOCO.scala

```diff
@@ -74,38 +96,104 @@ class RecordInsightsLOCO[T <: Model[T]]
 
   private lazy val featureInfo = OpVectorMetadata(getInputSchema()(in1.name)).getColumnHistory().map(_.toJson(false))
 
-  override def transformFn: OPVector => TextMap = (features) => {
-    val baseScore = modelApply(labelDummy, features).score
-    val maxHeap = PriorityQueue.empty(MinScore)
+  private def computeDiffs(i: Int, oldInd: Int, featureArray: Array[(Int, Double)], featureSize: Int,
+    baseScore: Array[Double]): Array[Double] = {
+    featureArray.update(i, (oldInd, 0))
+    val score = modelApply(labelDummy, OPVector(Vectors.sparse(featureSize, featureArray))).score
+    val diffs = baseScore.zip(score).map { case (b, s) => b - s }
+    diffs
+  }
+
+  private def returnTopPosNeg(filledSize: Int, featureArray: Array[(Int, Double)], featureSize: Int,
+    baseScore: Array[Double], k: Int, indexToExamine: Int): Seq[(Int, Double, Array[Double])] = {
+    // Heap that will contain the top K positive LOCO values
+    val positiveMaxHeap = PriorityQueue.empty(MinScore)
+    // Heap that will contain the top K negative LOCO values
+    val negativeMaxHeap = PriorityQueue.empty(MaxScore)
+    // for each element of the feature vector != 0.0
+    // Size of positive heap
+    var positiveCount = 0
+    // Size of negative heap
+    var negativeCount = 0
+    var i = 0
+    while (i < filledSize) {
+      val (oldInd, oldVal) = featureArray(i)
+      val diffs = computeDiffs(i, oldInd, featureArray, featureSize, baseScore)
+      val max = diffs(indexToExamine)
+
+      if (max > 0.0) { // if positive LOCO then add it to positive heap
+        positiveMaxHeap.enqueue((i, max, diffs))
+        positiveCount += 1
+        if (positiveCount > k) { // remove the lowest element if the heap size goes from 5 to 6
+          positiveMaxHeap.dequeue()
+        }
+      } else if (max < 0.0) { // if negative LOCO then add it to negative heap
+        negativeMaxHeap.enqueue((i, max, diffs))
+        negativeCount += 1
+        if (negativeCount > k) { // remove the highest element if the heap size goes from 5 to 6
+          negativeMaxHeap.dequeue()
+        } // Not keeping LOCOs with value 0
+      }
+      featureArray.update(i, (oldInd, oldVal))
```

Comment:
would it make sense to move the update of the array back tot the old version into the compute diffs function?